### PR TITLE
Support global registry configuration

### PR DESCRIPTION
**Pull request by ahmed-mez**
### What does this PR do?

Support a global registry config.

### Motivation

Configure the container registry without depending on the image name and tag.

### Additional Notes

For backward compatibility, the tag and registry values are ignored in case Image.Name corresponds to a full image name.

### Describe your test plan

Example CR:
```
apiVersion: datadoghq.com/v1alpha1
kind: DatadogAgent
metadata:
  name: datadog
spec:
  registry: public.ecr.aws/datadog
...
  agent:
    image:
      name: "agent"
      tag: "latest"
  clusterAgent:
    replicas: 1
    image:
      name: "cluster-agent"
      tag: "latest"
    config:
      clusterChecksEnabled: true
  clusterChecksRunner:
    replicas: 1
    image:
      name: "agent"
      tag: "latest"
```

### CODE REVIEW

File: api/v1alpha1/datadogagent_default.go [hunk 1]
```diff
@@ -20,7 +20,9 @@ import (
 // default values
 const (
 	DefaultLogLevel                       string = "INFO"
-	defaultAgentImage                     string = "gcr.io/datadoghq/agent:latest"
+	defaultImageTag                       string = "latest"
```
Comment (clamoriniere) [inline api/v1alpha1/datadogagent_default.go:23]: we should start to use specific version: `1.12.0` for example

File: api/v1alpha1/datadogagent_types.go [hunk 2]
```diff
@@ -1145,11 +1151,18 @@ type DatadogAgentSpecClusterChecksRunnerSpec struct {
...
 	// Use "datadog/dogstatsd:latest" for Standalone Datadog Agent DogStatsD6
 	// Use "gcr.io/datadoghq/cluster-agent:latest" for Datadog Cluster Agent
+	// Use "agent" with the registry and tag configurations for <registry>/agent:<tag>
+	// Use "cluster-agent" with the registry and tag configurations for <registry>/cluster-agent:<tag>
 	Name string `json:"name"`
 
+	// Define the image version to use:
+	// To be used if the Name field doesn't correspond to a full image string.
+	// +optional
+	Tag string `json:"tag"`
```
Comment (clamoriniere) [inline api/v1alpha1/datadogagent_types.go:1164]: `omitempty` can be added.

```suggestion
	Tag string `json:"tag,omitempty"`
```

File: api/v1alpha1/datadogagent_types.go [hunk 3]
```diff
@@ -1145,11 +1151,18 @@ type DatadogAgentSpecClusterChecksRunnerSpec struct {
 // +k8s:openapi-gen=true
 type ImageConfig struct {
 	// Define the image to use:
-	// Use "gcr.io/datadoghq/agent:latest" for Datadog Agent 6
+	// Use "gcr.io/datadoghq/agent:latest" for Datadog Agent 7
 	// Use "datadog/dogstatsd:latest" for Standalone Datadog Agent DogStatsD6
 	// Use "gcr.io/datadoghq/cluster-agent:latest" for Datadog Cluster Agent
+	// Use "agent" with the registry and tag configurations for <registry>/agent:<tag>
+	// Use "cluster-agent" with the registry and tag configurations for <registry>/cluster-agent:<tag>
 	Name string `json:"name"`
```
Comment (clamoriniere) [inline api/v1alpha1/datadogagent_types.go:1159]: `omitempty` can be added.

```suggestion
	Name string `json:"name,omitempty"`
```

File: api/v1alpha1/datadogagent_types.go [hunk 4]
```diff
@@ -1145,11 +1151,18 @@ type DatadogAgentSpecClusterChecksRunnerSpec struct {
...
 
+	// Define the image version to use:
+	// To be used if the Name field doesn't correspond to a full image string.
+	// +optional
+	Tag string `json:"tag"`
+
```
Comment (clamoriniere) [inline api/v1alpha1/datadogagent_types.go:1165]: Open question: should we also provide the possibility to set if the user want to use the JMX image separately.

Something like:

```suggestion

	JMXSupportEnabled bool `json:"jmxSupportEnabled,omitempty"`
```

`if JMXSupportEnabled == true` the operator append to the end of the tag `-jmx`, and so the operator can still default properly the tag (7.28.0 for example)

**Review by clamoriniere — COMMENTED** [review]
Few idea to provide more options to the user.

File: api/v1alpha1/datadogagent_types.go (hunk 4 above)
Comment (ahmed-mez) [reply api/v1alpha1/datadogagent_types.go:1165]: agreed, this can be useful!

**Comment by vboulineau** [comment]
I will conflict a bit with https://github.com/DataDog/datadog-operator/pull/303 (see usage of `GetTagFromImageName`)

**Comment by ahmed-mez** [comment]
@vboulineau do you want to merge #303 first?

File: config/crd/bases/v1/datadoghq.com_datadogagents.yaml [hunk 5]
```diff
@@ -2881,8 +2887,10 @@ spec:
                               type: string
                           type: object
                         type: array
-                    required:
-                    - name
+                      tag:
+                        description: 'Define the image version to use: To be used
+                          if the Name field doesn''t correspond to a full image string.'
```
Comment (jtappa) [inline config/crd/bases/v1/datadoghq.com_datadogagents.yaml:2892]:
```suggestion
                          if the Name field does not correspond to a full image string.'
```

File: config/crd/bases/v1/datadoghq.com_datadogagents.yaml [hunk 6]
```diff
@@ -6377,8 +6391,10 @@ spec:
...
```
Comment (jtappa) [inline config/crd/bases/v1/datadoghq.com_datadogagents.yaml:6396]:
```suggestion
                          if the Name field does not correspond to a full image string.'
```

File: config/crd/bases/v1/datadoghq.com_datadogagents.yaml [hunk 7]
```diff
@@ -8935,8 +8957,10 @@ spec:
...
```
Comment (jtappa) [inline config/crd/bases/v1/datadoghq.com_datadogagents.yaml:8962]:
```suggestion
                          if the Name field does not correspond to a full image string.'
```

File: config/crd/bases/v1beta1/datadoghq.com_datadogagents.yaml [hunk 8]
```diff
@@ -2781,8 +2786,10 @@ spec:
                             type: string
                         type: object
                       type: array
-                  required:
-                  - name
+                    tag:
+                      description: 'Define the image version to use: To be used if
+                        the Name field doesn''t correspond to a full image string.'
```
Comment (jtappa) [inline config/crd/bases/v1beta1/datadoghq.com_datadogagents.yaml:2791]:
```suggestion
                        the Name field does not correspond to a full image string.'
```

File: config/crd/bases/v1beta1/datadoghq.com_datadogagents.yaml [hunk 9]
```diff
@@ -6158,8 +6170,10 @@ spec:
...
```
Comment (jtappa) [inline config/crd/bases/v1beta1/datadoghq.com_datadogagents.yaml:6175]:
```suggestion
                        the Name field does not correspond to a full image string.'
```

File: config/crd/bases/v1beta1/datadoghq.com_datadogagents.yaml [hunk 10]
```diff
@@ -8632,8 +8651,10 @@ spec:
...
```
Comment (jtappa) [inline config/crd/bases/v1beta1/datadoghq.com_datadogagents.yaml:8656]:
```suggestion
                        the Name field does not correspond to a full image string.'
```

**Review by jtappa — CHANGES_REQUESTED** [review]

**Review by vboulineau — APPROVED** [review]

**Pull request merged**